### PR TITLE
Fix: avoid embedding the user input into the error response.

### LINKED ISSUE #2107: Error message contains whole payload when the type is not correct
**Describe the bug**
I tried to send (wrongly) a string using `POST /indexes/x/documents`. The returned error message from MeiliSearch is as follows: 
`MeiliSearchApiError: The json payload provided is malformed. "whole payload"`.

In my case the `whole payload` is 3million character long. So my error message looks like this:

Screenshot 2022-01-24 at 13 39 54

(ignore the popup from vscode)

**To Reproduce**
1. Add a string instead of a document with content-type json
2. It throws an error containing the whole string

**Expected behavior**
Maybe a more explicit error `The json payload provided is malformed, expected an array but go a string instead`

**Screenshots**
see above

**MeiliSearch version:** v0.25.1

### PR DESCRIPTION
**Pull request by CNLHC**
# Pull Request

## What does this PR do?
Fix #2107. 

The problem is meilisearch embeds the user input to the error message. 

The reason for this problem is `milli` throws a `serde_json: Error` whose `Display` implementation will do this embedding.  

I tried to solve this problem in this PR by manually implementing the `Display` trait for `DocumentFormatError` instead of deriving automatically.

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: meilisearch-lib/src/document_formats.rs [hunk 1]
```diff
@@ -23,16 +24,27 @@ impl fmt::Display for PayloadType {
...
+        match self {
+            Self::Internal(e) => write!(f, "An internal error has occurred. `{}`.", e),
+            Self::MalformedPayload(me, b) => match me.borrow() {
+                milli::documents::Error::JsonError(se) => write!(
+                    f,
+                    "The `{}` payload provided is malformed. `Couldn't serialize document value. at line {} column {}`",
+                    b,se.line(),se.column()
+                ),
+                _ => write!(f, "The `{}` payload provided is malformed. `{}`.", b, me),
+            },
```
Comment (Kerollmops) [inline meilisearch-lib/src/document_formats.rs:43]:
```suggestion
            Self::Internal(e) => write!(f, "An internal error has occurred: `{}`.", e),
            Self::MalformedPayload(me, b) => match me.borrow() {
                milli::documents::Error::JsonError(se) => write!(
                    f,
                    "The `{}` payload provided is malformed. `Couldn't serialize document value at line {} column {}`",
                    b, se.line(), se.column()
                ),
                _ => write!(f, "The `{}` payload provided is malformed: `{}`.", b, me),
            },
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Thank you very much for your PR!
Apart from some esthetic changes, it's good for me.

File: meilisearch-lib/src/document_formats.rs [hunk 2]
```diff
@@ -23,16 +24,27 @@ impl fmt::Display for PayloadType {
...
 pub enum DocumentFormatError {
-    #[error("An internal error has occurred. `{0}`.")]
     Internal(Box<dyn std::error::Error + Send + Sync + 'static>),
-    #[error("The `{1}` payload provided is malformed. `{0}`.")]
-    MalformedPayload(
-        Box<dyn std::error::Error + Send + Sync + 'static>,
-        PayloadType,
-    ),
+    MalformedPayload(Box<milli::documents::Error>, PayloadType),
+}
```
Comment (Kerollmops) [inline meilisearch-lib/src/document_formats.rs:31]:
```suggestion
}

```

File: meilisearch-lib/src/document_formats.rs [hunk 3]
```diff
@@ -23,16 +24,27 @@ impl fmt::Display for PayloadType {
...
+                    f,
+                    "The `{}` payload provided is malformed. `Couldn't serialize document value at line {} column {}`",
+                    b, se.line(), se.column()
+                ),
+                _ => write!(f, "The `{}` payload provided is malformed: `{}`.", b, me),
+            },
+        }
+    }
 }
+impl std::error::Error for DocumentFormatError {}
```
Comment (Kerollmops) [inline meilisearch-lib/src/document_formats.rs:47]:
```suggestion

impl std::error::Error for DocumentFormatError {}
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]
I had some thoughts about your PR and think it removes too much information about the error, users will be lost if an error occurs and only the line and column are specified.

What do you think about ellipsing the message? By using something like 50 chars at the start and 50 chars at the end:
```
The `json` payload provided is malformed. `Couldn't serialize document value: invalid type: string \"my very ver...very very long body here\", expected a documents, or a sequence of documents. at line 1 column 7`.
```

**Comment by CNLHC** [comment]
Sorry for the late response. I will submit a patch to  follow the suggestion of @Kerollmops today.

**Comment by curquiza** [comment]
@CNLHC thanks a lot for these changes 🙏 
Looks like there is a tiny issue to fix with our clippy CI before Kero reviews it :)

File: meilisearch-http/tests/documents/add_documents.rs [hunk 4]
```diff
@@ -307,6 +308,71 @@ async fn error_add_malformed_json_documents() {
...
+    let long = String::from_utf8(
+        "0123456789"
+            .as_bytes()
+            .iter()
+            .cycle()
+            .cloned()
+            .take(100)
+            .collect_vec(),
+    )
+    .unwrap();
```
Comment (Kerollmops) [inline meilisearch-http/tests/documents/add_documents.rs:324]:
```suggestion
    let long = "0123456789".repeat(10);
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Thank you very much for your contribution ❤️
However, could you modify the tests according to the changes I proposed with the ellipsing algorithm? And could you stash everything into one final commit when we are done, please?

File: meilisearch-lib/src/document_formats.rs [hunk 5]
```diff
@@ -23,17 +24,47 @@ impl fmt::Display for PayloadType {
...
+                        } else {
+                            serde_msg = String::from("");
+                        }
+                    }
+
+                    write!(
+                    f,
+                    "The `{}` payload provided is malformed. `Couldn't serialize document value: {}`.",
+                    b,serde_msg
+                )
```
[the remark this answers is not in the document]
Comment (CNLHC) [reply meilisearch-lib/src/document_formats.rs]: In some cases, this method will lead to weird error messages. For example, if the raw message is 
```
The `json` payload provided is malformed. `Couldn't serialize document value: invalid type: string "0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789", expected a documents, or a sequence of documents. at line 1 column 102`.
```

Using your method, ellipsing the string without parsing the context will  generate

```
The `json` payload provided is malformed. `Couldn't serialize document value: invalid type: string \"0123456789012345678901234567..., or a sequence of documents. at line 1 column 102`.
```

Some meaningful information(The `expected a documents`)  is truncated.

[the remark this answers is not in the document]
Comment (CNLHC) [reply meilisearch-lib/src/document_formats.rs]: The meaningful suffix length cannot be determined. But in most cases, its length is around `75` which is 
`", expected a documents, or a sequence of documents. at line <LINE_NUMBER> column <COLUMN_NUMBER>`

So I suggest to change  `50..serde_msg.len() - 50` to 50..serde_msg.len() - 85`

File: meilisearch-http/tests/documents/add_documents.rs [hunk 6]
```diff
@@ -307,6 +307,62 @@ async fn error_add_malformed_json_documents() {
...
+    let req = test::TestRequest::put()
+        .uri("/indexes/dog/documents")
+        .set_payload(document)
+        .insert_header(("content-type", "application/json"))
+        .to_request();
+    let res = test::call_service(&app, req).await;
+    let body = test::read_body(res).await;
+    dbg!(&body);
+    let response: Value = serde_json::from_slice(&body).unwrap_or_default();
+    dbg!(&response);
```
Comment (Kerollmops) [inline meilisearch-http/tests/documents/add_documents.rs:326]:
```suggestion
    let response: Value = serde_json::from_slice(&body).unwrap_or_default();
```

File: meilisearch-http/tests/documents/add_documents.rs [hunk 7]
```diff
@@ -307,6 +307,62 @@ async fn error_add_malformed_json_documents() {
...
```
Comment (Kerollmops) [inline meilisearch-http/tests/documents/add_documents.rs:352]:
```suggestion
    let response: Value = serde_json::from_slice(&body).unwrap_or_default();
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Thank you, that's better and clever. Could you remove the `dbg!` in the test and I'll merge after that, please?
Once you have done so, could you squash everything in one commit and rebase on main, please?

File: meilisearch-http/tests/documents/add_documents.rs (hunk 7 above)
Comment (CNLHC) [reply meilisearch-http/tests/documents/add_documents.rs]: Oh, It's my fault. I should remove it.

**Review by Kerollmops — APPROVED** [review]
Thank you very much for your work @CNLHC, merging now!
bors merge

**Pull request merged**